homeassistant/components/automation/zone.py

```diff
@@ -46,6 +47,23 @@ def zone_automation_listener(entity, from_s, to_s):
             from_match = False
         to_match = condition.zone(hass, zone_state, to_s)
 
+        #
+        #  If zone is "zone.*", trigger automation when entering or leaving
+        #
+        if zone_entity_id == MATCH_ALL and from_s.state != to_s.state:
```

Comment:
I have gone over this code several times trying to get this complex if/then/else logic to combine the two scenarios, but it gets SO ugly.  I am happy for suggestions.  It was already so complex that the from_match/to_match was externalised with condition.zone.  When adding a check for zone * AND moving from a 'non-zone' to a 'zone' scenario, it makes the code very unreadable.  Again, I'm happy to see an example if/then/else which is clear and handles the two scenarios.